src/blas/impl/KokkosBlas3_gemm_impl.hpp

```diff
@@ -0,0 +1,506 @@
+  typedef Kokkos::Details::ArithTraits<value_type>     ATV;
+
+  KOKKOS_INLINE_FUNCTION
+  static void update(const TeamHandle& team, const value_type& beta , const ViewType& A,
+                                             const value_type& alpha, const ViewTypeScratch& A_scr,
+                                             const int& offset_i, const int& offset_j) {
+    if(offset_i + blockDim_i <= A.extent(0) && offset_j + blockDim_j <= A.extent(1)) {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        if(beta == ATV::zero()) {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = alpha * A_scr(i,j);
+          });
+        } else {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = beta * A(idx_i,idx_j) + alpha * A_scr(i,j);
+          });
+        }
+      });
+    } else {
+      const int range_i = offset_i + blockDim_i <= A.extent(0)?blockDim_i:A.extent(0)%blockDim_i;
+      const int range_j = offset_j + blockDim_j <= A.extent(1)?blockDim_j:A.extent(1)%blockDim_j;
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,range_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        if(beta == ATV::zero()) {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,range_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = alpha * A_scr(i,j);
+          });
+        } else {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,range_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = beta * A(idx_i,idx_j) + alpha * A_scr(i,j);
+          });
+
+        }
+      });
+    }
+  }
+};
+
+// Compute a single A block 8 B block, also do an in-place no-additional blocking team GEMM
+template<class TeamHandle, class ViewTypeA, class ViewTypeB, class ViewTypeC>
+KOKKOS_INLINE_FUNCTION
+void impl_team_gemm_block(const TeamHandle& team, const ViewTypeC& C, const ViewTypeA& A, const ViewTypeB& B) {
+  typedef typename ViewTypeC::non_const_value_type ScalarC;
+  const int blockA0 = A.extent(0);
+  const int blockA1 = A.extent(1);
+  const int blockB1 = B.extent(1);
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockA0), [&] (const int i) {
+#if defined(__CUDA_ARCH__) || !defined(KOKKOS_ENABLE_OPENMP)
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockB1/4), [&] (const int B_j) {
+#else
+    #pragma omp simd
+    for(int B_j=0; B_j<blockB1/4; B_j++) {
+#endif
+      ScalarC C_ij0 = 0;
+      ScalarC C_ij1 = 0;
+      ScalarC C_ij2 = 0;
+      ScalarC C_ij3 = 0;
+      for(int j = 0; j < blockA1; j++) {
+        ScalarC A_ij = A(i,j);
+        C_ij0 += A_ij*B(j,B_j);
```

Comment:
Why do we need to cast to ``ScalarC`` ? This means that the function does not allow mixed precision ?

Reply:
We do not cast to ScalarC. We accumulate in ScalarC, which I believe is the right thing to do. A and B can still be a different precision Level.

Reply:
I mean ``ScalarC A_ij = A(i,j)``

Reply:
Hm yeah, but I am not sure that it matters. If this is converted into fma anyway I would assume all three arguments will be converted to ScalarC. In some sense I feel we may actually want to explicitly cast B as well. What do you think?

Reply:
Probably possible case is ScalarA and ScalarB are float and and ScalarC is double. Then, we compute low precision multiplication and addition (?). I agree that it is not important. 

Or other possibility may come from complex arithmetics ? A and B are real matrix and C is complex ?

Reply:
The complex case is interesting, but not sure whether it is relevant. C, complex and one of A and B being complex while the other one is real, that yes but both A and B real, seems weird. The A and B float case I think it would be better to convert everything to double to use fma. I dont' think you would gain anything from not using fma, in order to do the multiplication in float and then the addition in double. It would probably drive vectorization crazy as well since the vector length of the ops doesn't match up.

Reply:
Sierra uses mixed-precision sparse mat-vec, with float matrix and double vectors.  

If I were focusing on correctness and maximum generality, I would actually prefer either not doing the explicit cast, or using `std::common_type`.  I would also consider an option to do intermediate sums in extended precision.  On the other hand, I'm glad that @crtrott is writing this ;-) .